things to fix before resubmitting to CRAN

idavydov opened the pull request:
While working on this I removed `R/randomization.R`; we have several unexported functions which were responsible for most of the warnings. We can later recover the file if needed.

- [x] proofread DESCRIPTION
- [x] If there are references describing the methods in your package, pleas add these in the description 
- [x] examples for unexported functions
```
examples or export these functions. -> Warning: Examples for unexported
function
   random_score_variances() in:
      randomize.Rd
```
- [x] \dontrun{} should only be used if the example really cannot be executed
- [x] Instead of cat() rather use message()/warning() or if(verbose)cat(..) -> R/randomization.R
- [x] do not set a seed to a specific number within a function. -> R/optimize.R
- [x] do not modify the global environment (e.g. by using <<-) tests/testthat/test_optimize_design_simple_shuffle.R
- [x] add @accio

idavydov commented:
> Please do not set a seed to a specific number within a function

[Here's](https://github.com/bedapub/designit/blob/main/R/optimize.R#L201) the relevant code:
``` r
  # based on https://stat.ethz.ch/pipermail/r-help/2007-September/141717.html
  if (!exists(".Random.seed")) stats::runif(1)
  trace$seed <- list(.Random.seed)
```

I personally do not think this qualifies as setting a seed; however I agree that this is not an idiomatic code.

@klmr @banfai @ingitwetrust what's your suggestion? 

the path of least resistance is just removing this.

klmr commented:
Without prior knowledge of the structure it isn’t obvious to me what the purpose of `trace$seed` is.

If the purpose is reproducibility, it should be sufficient to replace the code in question with

```
trace$seed <- list(get0(".Random.seed", .GlobalEnv))
```

This will assign `NULL` if the seed isn’t yet set. However, this now requires special handling when using the stored seed, since `NULL` is an invalid random seed.

And note that it’s not sufficient to store the random seed to get full reproducibility, we also need to store the RNG kind and version.

klmr commented:
Oops, the previous comment is nonsense: storing `NULL` is obviously *not* sufficient to get reproducibility. We do actually need to seed the RNG under these circumstances; but we should use the proper seeding function for that instead of using the side-effect from `runif(1)`.

There doesn’t seem to be an idiomatic way of seeding the RNG with a good random number using `set.seed()`. However, the following should do the trick, and I would call it idiomatic based on the examples from the documentation:

```r
if (!exists(".Random.seed")) {
  do.call(RNGkind, as.list(RNGkind()))
}
```

(Actually the documentation is incorrect because it implies that `RNGkind()` should also work and select the default RNG, but that is not the case.)

idavydov commented:
thanks, @klmr. I updated the code to save RNG kind and I do not use side-effects of `runif()`.
I also extensively commented this fragment and added tests.

could you please review?

idavydov commented on tests/testthat/test-save-random-seed.R:
```diff
@@ -0,0 +1,46 @@
+
+test_that("random seed is saved when no seed is set", {
+  # unset random seed
+  rm(.Random.seed, envir = .GlobalEnv)
```
hi @klmr ,
in this file I'm testing that random seed and rng kind is saved in `bc$trace` both when:
- random seed is not set yet
- random seed is already set (that's easy)

I'm using `rm(.Random.seed, envir = .GlobalEnv)` to simulate a situation when random seed wasn't set


Do you think the reviewer would argue that this is bad code since it modifies the global environment?

Do you have a better suggestion how to test this?

- `set.seed(NULL)` doesn't seem to remove `.Random.seed`
- I couldn't find a way to start the test in a clean environment.
- if I use empty env, nothing works (since e.g. `{` is undefined)
- something with `local_mocked_bindings()`?

banfai replied:
```suggestion
  if (exists(".Random.seed")) {
    random_seed <- .Random.seed
    on.exit(.GlobalEnv$.Random.seed <- random_seed)
    rm(.Random.seed, envir = .GlobalEnv)
    testthat::expect_false(exists(".Random.seed"))
  }
  
  ```

idavydov commented:
Facilitate experiment design...

idavydov commented on R/optimize.R:
```diff
@@ -197,9 +197,16 @@ optimize_design <- function(batch_container, samples = NULL,
+    # .Random.seed in the global environment.
+    # This doesn't affect the way random seed would have been set in
+    # downstream functions such as `sample()`.
+    do.call(RNGkind, as.list(RNGkind()))
```
set.seed(NULL)

idavydov replied:
```suggestion
    set.seed(NULL)
```

idavydov commented on tests/testthat/test-save-random-seed.R:
```diff
@@ -0,0 +1,46 @@
+scoring_f <- function(...) rnorm(1)
+
+test_that("random seed is saved when no seed is set", {
+  # unset random seed
```
```
if (exists(..)) {
  random_seed <- .Random.seed
  on.exit(restore)
  rm(.Random.seed, envir = .GlobalEnv)
}
```

idavydov commented:
hi @klmr, I think it's ready for review now.

@ingitwetrust regarding design of experiment, I forgot but it's in the title. so should be quite findable.
